### PR TITLE
Recommended Articles Section (bottom of blog article page)

### LINKED ISSUE #246: Add Recommended Articles Section to Blog Article Page
**Description**
Currently, the page that renders a blog article is missing the recommended blog section on the bottom.

https://www.figma.com/file/k6TNqd4V49TsJlsvaZKnQV/International-Student-Mentorship-Program?node-id=303%3A469

- [x] I've added the proper component `comp/*` label + other labels
- [x] I've added any context to where the decision was made (Basecamp links...etc.)

### PR DESCRIPTION
**Pull request by vickyadebona**
## Proposed changes
Added the "Recommended Articles" section below the blog article according to [Figma](https://www.figma.com/file/k6TNqd4V49TsJlsvaZKnQV/International-Student-Mentorship-Program?node-id=303%3A469)

## Types of changes

- [x] I've labeled this PR with appropriate labels (`comp/*` component, bug/feature, documentation and other less important ones).

## Checklist
_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging your code._

- [x] I've linked the proper Github Issue(s) to my PR
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests (unit, integration...etc.) that prove my fix is effective or that my change works
- [ ] I have added or updated necessary documentation (if appropriate). You can search, for instance using the [`silver searcher`](https://github.com/ggreer/the_silver_searcher) `ag <STRING>` to find all usages in the repository.
- [ ] I have notified the team of any breaking or important changes

closes #246

### CODE REVIEW

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 1]
```diff
@@ -408,6 +530,26 @@ class Blogpost extends Component {
...
+                    return (
+                      <BlogCardWrapper key={index}>
+                        <BlogCard sourcePage='blogpost' blogPost={article} position={index}/>
+                      </BlogCardWrapper>
+                    )
+                  }
+                })
+              }
+            </RecommendedArticleContainer>
+          </RecommendedArticles>
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:552]: This feels like it should be a separate component.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 2]
```diff
@@ -327,6 +405,50 @@ class Blogpost extends Component {
...
+      // Sorting backwards --> newest to oldest
+      if (blogA > blogB) {
+        return -1;
+      } 
+      if (blogA < blogB) {
+        return 1;
+      }
+    });
+    return sortedBlogs;
+  }
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:450]: If you do end up refactoring, then we can move this logic too.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 3]
```diff
@@ -408,6 +530,26 @@ class Blogpost extends Component {
...
+          <RecommendedArticles>
+            <RecommendedArticlesHeader>
+              <RecommendedArticleBorder />
+              Recommended Articles
+              <RecommendedArticleBorder />
+            </RecommendedArticlesHeader>
+            <RecommendedArticleContainer>
+              {
+                this.state.recommendedArticles.map((article, index) => {
+                  if (article.id !== this.state.blogPostContent.id) {
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:542]: Can we add this check to where we actually populate the `recommendedArticles` state? Also a way to guarantee we actually get three articles, if there are three available.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 4]
```diff
@@ -327,6 +405,50 @@ class Blogpost extends Component {
...
+    } catch(error) {
+      console.log('could not get recommended articles: ', error);
+    }
+  };
+
+  sortRecommendedBlogs(blogs) {
+    let sortedBlogs = blogs;
+    sortedBlogs.sort((a, b) => {
+      const blogA = new Date(a.publish_at);
+      const blogB = new Date(b.publish_at);
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:440]: nit: probably `dateA` and `dateB` are better variable names.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 5]
```diff
@@ -327,6 +405,50 @@ class Blogpost extends Component {
...
+      // Sorting backwards --> newest to oldest
+      if (blogA > blogB) {
+        return -1;
+      } 
+      if (blogA < blogB) {
+        return 1;
+      }
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:447]: What if they are equal?

Also, could just shorten this to `blogA - blogB` (or maybe supposed to be `blogB - blogA` - point is, can shorten this).

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 6]
```diff
@@ -265,49 +324,68 @@ class Blogpost extends Component {
...
-          });
-          logContentView(result);
+    this.populateBlogpostData();
+
+    // Listening for a change to the blog id, 
+    // for when a new blog needs to be loaded
+    this.historyListener = this.props.history.listen(location => {
+      const pathPieces = location.pathname.split('/');
+      this.setState({ blogPostContentId: pathPieces[2]});
+      window.location.reload();
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:340]: Just wondering, do we need this? I'm not 100% sure of what's going on here - but seems overkill to reload everything? Shouldn't everything reload because we are updating the state?

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 7]
```diff
@@ -396,7 +518,7 @@ class Blogpost extends Component {
                 {moment(this.state.blogPostContent.publish_at).format(
                   'MMMM DD, YYYY'
                 )}{' '}
-                {bio}
+                by {this.state.blogPostContent.author_display_name}
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:521]: Maybe one thing to clean up is to decide if we even need `this.getBlogData()` anymore? Would rather use that method for everything, or totally get rid of it if it doesn't serve ag ood purpose.

**Comment by vickyadebona** [comment]
@dbtung this is ready for you to look at again

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 8]
```diff
@@ -265,68 +275,65 @@ class Blogpost extends Component {
...
+    this.populateBlogpostData();
+
+    // Listening for a change to the blog id, 
+    // for when a new blog needs to be loaded
+    this.historyListener = this.props.history.listen(location => {
+      const pathPieces = location.pathname.split('/');
+      this.setState({ blogPostContentId: pathPieces[2]}, () => {
+        this.populateBlogpostData();
+      });
+    });
```
Comment (JonSEng) [inline frontend/src/pages/Blogs/Blogpost.jsx:293]: Hi @vickyadebona for my context and learning, when will a blog id change?

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 9]
```diff
@@ -408,6 +419,14 @@ class Blogpost extends Component {
             Get connected with a mentor today
             <ApplyNowBtn to="/apply">APPLY NOW</ApplyNowBtn>
           </GetConnected>
+          <RecommendedArticles 
+            topic={
+              this.state.blogPostData 
+              && this.state.blogPostData.topic_set.length > 0 
+              && this.state.blogPostData.topic_set[0].title
+            }
```
Comment (JonSEng) [inline frontend/src/pages/Blogs/Blogpost.jsx:427]: Hi @vickyadebona is there a difference between the `topic` here and the one [here](https://github.com/International-Mentorship-ISMP/ismp/pull/351/files#diff-2cb9127405b37d2e3f2c09d3ac36c73bR390)?

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 10]
```diff
@@ -379,7 +386,11 @@ class Blogpost extends Component {
         </ErrorPageContainer>
       );
     } else {
-      const { bio, topic, headerMedia } = this.getBlogData();
+      // Show either blog type or topic category
```
Comment (JonSEng) [inline frontend/src/pages/Blogs/Blogpost.jsx:389]: For my context, what about our design makes it so we have to do this search? It seems counter-intuitive for me as I would expect we would always look up the blog type or post's topic, but not check both unless the `blogPostData.topic_set` isn't always defined for some reason. If that were the case, would it be because we have malformed data we retrieved?

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 8 above)
Comment (vickyadebona) [reply frontend/src/pages/Blogs/Blogpost.jsx:293]: @JonSEng the ID changes when one of the blog cards for a recommended articles is pressed. The blog article component would then need to re-render with the new blog content.

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 9 above)
Comment (vickyadebona) [reply frontend/src/pages/Blogs/Blogpost.jsx:427]: The one listed about is the name that is used for querying blogs (when I make the api call to get recommended/related blogposts), whereas the other one has the full topic name displayed at the top of the blog page (`.title` versus `.display_text`)

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 10 above)
Comment (vickyadebona) [reply frontend/src/pages/Blogs/Blogpost.jsx:389]: I'm not exactly sure how it works, but for some reason, it's two endpoints combined into one in order to get the associated blog information. I think @douglaschan32167 would know more about how the blog posts are structured. But otherwise, yeah, I had some issues with React trying to render data that isn't defined yet.

File: frontend/src/components/RecommendedArticles/RecommendedArticles.jsx [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+                articles = articles.filter((article) => {
+                    return article.id !== currentArticle;
+                });
+
+                // Only picking top three
+                articles = articles.slice(0, 3)
+                this.setState({ articles });
+            });
+        } catch(error) {
+            console.log('could not get recommended articles: ', error);
```
[the remark this answers is not in the document]
Comment (vickyadebona) [reply frontend/src/components/RecommendedArticles/RecommendedArticles.jsx]: @JonSEng I'm not sure how I would configure the `debug` thing. I tried looking it up earlier, and it seems like it's a `debugger` keyword in javascript that runs the chrome debugger to help catch and step through the error.

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 10 above)
Comment (douglaschan32167) [reply frontend/src/pages/Blogs/Blogpost.jsx:390]: Hi @JonSEng ,
Not sure why it was designed like this in the first place. topic_set may be an empty list (if we don't specify a topic when creating a blogpost) but showing the blog type in its place is a little non-intuitive. For reference, blog type is either "blogpost" or "webinar"

**Comment by vickyadebona** [comment]
@dbtung if you have time, can you take another look? I'll need your approval of this PR in order for it to be merged.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 12]
```diff
@@ -265,68 +275,66 @@ class Blogpost extends Component {
...
-    };
-    populateBlogpostData();
+    this.populateBlogpostData();
+
+    // Listening for a change to the blog id, 
+    // for when a new blog needs to be loaded
+    // when a recommended blog article is pressed
+    this.historyListener = this.props.history.listen(location => {
+      const pathPieces = location.pathname.split('/');
+      this.setState({ blogPostContentId: pathPieces[2]}, () => {
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:291]: nit: Need extra space after `pathPieces[2]`

Also, why do we need to put `this.populateBlogpostData()` here? Feels like it belongs in `componentDidUpdate`? SInce that will be called, since we are updating the state?

File: frontend/src/layout/Navigation/Navigation.jsx [hunk 13]
```diff
@@ -244,10 +244,10 @@ const Nav = ({ mobile, history }) => {
         )}
         {mobile && (
           <div>
-            <NavLink to="/about" active={currentPath === '/about'}>
+            <NavLink to="/about" active={currentPath === '/about' ? 1 : 0}>
```
[the remark this answers is not in the document]
Comment (vickyadebona) [reply frontend/src/layout/Navigation/Navigation.jsx:247]: @dbtung I was getting some console errors for not having a boolean value instead of an int or string being passed into the `active` prop, so I opted for 1 and 0. If not for the errors, I would've left it as is.

File: frontend/src/components/RecommendedArticles/RecommendedArticles.jsx [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        });
+        return sortedArticles;
+    }
+
+    getRecommendedBlogArticles() {
+        const { topic, currentArticle } = this.props;
+        try {
+            // forming request url for articles
+            const request = topic ? `blogpostcontent/?topic=${topic}` : 'blogpostcontent/';
+            let articles = [];
```
[the remark this answers is not in the document]
Comment (vickyadebona) [reply frontend/src/components/RecommendedArticles/RecommendedArticles.jsx:95]: I thought it would look cleaner, but I can move it inside

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 8 above)
Comment (JonSEng) [reply frontend/src/pages/Blogs/Blogpost.jsx]: Got it thanks for explaining this.

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 10 above)
Comment (JonSEng) [reply frontend/src/pages/Blogs/Blogpost.jsx:393]: > If it's not associated with any, then topic_set is an empty list.

I think this is just a consideration for later.
Can we enforce the invariant that a post should have at least one topic? It doesn't make sense to me that a blog can have no topic.

Comment (douglaschan32167) [reply frontend/src/pages/Blogs/Blogpost.jsx:393]: Sorry, I realize it's not very well documented.
Topic is the table that contains the topics that will be shown in the program page. Because we're highlighting that, we want to be a be able to filter by those specifically. It will be a set that only the web people can add to. It's a many-to-many relationship, which is why it can be an empty set. I suppose we could add a "general" topic. I would need to think about it more. That may be a frontend-only thing.

>If it's a list of topics per blogcontent, could we have the head be the primary topic and we always take that one?
isn't that what we're doing here? Index 0? Or does head mean something else?

Comment (vickyadebona) [reply frontend/src/pages/Blogs/Blogpost.jsx:393]: @douglaschan32167 @JonSEng yeah, the logic is set so that the topic chosen is always the first one (index 0).

Comment (JonSEng) [reply frontend/src/pages/Blogs/Blogpost.jsx:393]: > Topic is the table that contains the topics that will be shown in the program page

I see I guess I was confused on the definition of `topic`.
So if I'm understanding this correctly:

1. `topic` refers to what's on the program page . It seems like we only have 4 topics right now 
Screenshot 2020-07-12 11 33 44

2. `type` is an enum that is either `blogpost` or `webinar`. So in the below case, this blog post has an empty `topic_set` so we default to the `type` which in this case is `webinar`.
Screenshot 2020-07-12 11 39 14


Is that right @dougchansolute ?
